tests/fe/fe_abf_gradient_divergence_theorem.cc

```diff
@@ -0,0 +1,168 @@
+#include <deal.II/fe/fe_nedelec.h>
+#include <deal.II/fe/fe_abf.h>
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_system.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_q1.h>
+
+#include <sstream>
+#include <fstream>
+
+
+
+template<int dim>
+Tensor<1,dim> ones ()
+{
+	Tensor<1,dim> result;
+	for(unsigned int i=0; i<dim; ++i)
+		result[i] = 1.0;
+	return result;
+}
+
+template<int dim>
+void test (const Triangulation<dim> &tr,
+		const FiniteElement<dim> &fe,
+		const double tolerance)
+{
+	DoFHandler<dim> dof(tr);
+	dof.distribute_dofs(fe);
+
+	std::stringstream ss;
+
+	deallog << "FE=" << fe.get_name() << std::endl;
+
+	const QGauss<dim> quadrature(4);
+	FEValues<dim> fe_values (fe, quadrature, update_gradients
+			| update_quadrature_points
+			| update_JxW_values);
+
+	const QGauss<dim-1> face_quadrature(4);
+	FEFaceValues<dim> fe_face_values (fe, face_quadrature,
+			update_values
+			| update_quadrature_points
+			| update_normal_vectors
+			| update_JxW_values);
+
+	for(typename DoFHandler<dim>::active_cell_iterator cell = dof.begin_active();
+			cell != dof.end();
+			++cell)
+	{
+		fe_values.reinit (cell);
+
+		deallog << "Cell nodes:" << std::endl;
+		for(unsigned int i=0; i<GeometryInfo<dim>::vertices_per_cell; ++i)
+			deallog << i << ": (" << cell->vertex(i) << ")" << std::endl;
+
+		bool cell_ok = true;
+
+		for (unsigned int c=0; c<fe.n_components(); ++c)
+		{
+			FEValuesExtractors::Scalar single_component (c);
+
+			for (unsigned int i=0; i<fe_values.dofs_per_cell; ++i)
+			{
+				ss << "component=" << c
+						<< ", dof=" << i
+						<< std::endl;
+
+				Tensor<1,dim> bulk_integral;
+				for (unsigned int q=0; q<fe_values.n_quadrature_points; ++q)
+				{
+					bulk_integral += fe_values[single_component].gradient(i,q) * fe_values.JxW(q);
+				}
+
+				Tensor<1,dim> boundary_integral;
+				for (unsigned int face=0; face<GeometryInfo<dim>::faces_per_cell; ++face)
+				{
+					fe_face_values.reinit(cell,face);
+					for (unsigned int q=0; q<fe_face_values.n_quadrature_points; ++q)
+					{
+						boundary_integral += fe_face_values[single_component].value (i,q)
+													* fe_face_values.normal_vector(q)
+													* fe_face_values.JxW(q);
+					}
+				}
+
+				if ( (bulk_integral-boundary_integral).norm_square() > tolerance )
+				{
+					deallog << "Failed:" << std::endl;
+					deallog << ss.str() << std::endl;
+					deallog << "    bulk integral=" << bulk_integral << std::endl;
+					deallog << "boundary integral=" << boundary_integral << std::endl;
+					deallog << "Error! difference between bulk and surface integrals is greater than " << tolerance << "!\n\n" << std::endl;
+					cell_ok = false;
+				}
+
+				ss.str("");
+			}
+		}
+
+		deallog << (cell_ok? "OK: cell bulk and boundary integrals match...\n" : "Failed divergence test...\n") << std::endl;
+	}
+}
+
+
+
+template<int dim>
+void test_hyper_ball(const double tolerance)
+{
+	Triangulation<dim> tr;
+	GridGenerator::hyper_cube (tr);
```

Comment:
I guess this was on purpose -- just leave a comment in the testcase why you chose a rectangle here, rather than the ball as in the other testcases.